### PR TITLE
Create makefile to generate BTF files

### PR DESCRIPTION
**Pull request by mauriciovasquezbernal**
Remove the tools/btfgen.sh script and replace it by a Makefile as done in https://github.com/iovisor/bcc/pull/3889. The advantage of this change is that we don't need to worry about updating the btfgen.sh script anymore.

### CODE REVIEW

**Comment by mauriciovasquezbernal** [comment]
How did you try to build it? I'm using `make ENABLE_BTFGEN=true gadget-default-container` and seems to work fine on my machine.

**Comment by mauriciovasquezbernal** [comment]
Could you please describe the exact commands you're using? The following ones work fine to me:

```
$ make ENABLE_BTFGEN=true gadget-default-container
# or 
$ docker build -f gadget-default.Dockerfile --build-arg ENABLE_BTFGEN=true .
```

**Comment by mauriciovasquezbernal** [comment]
@eiffel-fl I think I found the issue. Could you please check that there are not old .o files on the bpf folders of IG on your machine?

```
$ ls pkg/gadgets/*/tracer/bpf/*.o
```

For some reason make is assuming it need to update those files compiling the .c files using cc, and of course it's failing because cc is not able to compile bpf code.

**Review by alban — CHANGES_REQUESTED** [review]

File: Makefile.btfgen [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+# TODO (filter by architecture too when implemented)
+BPF_O_FILES = $(shell find pkg/gadgets -type f -name '*.o')
+
+ifndef BTFHUB_ARCHIVE
+$(error BTFHUB_ARCHIVE is undefined)
+endif
+
+ifndef ARCH
+$(error ARCH is undefined)
+endif
```
Comment (alban) [inline Makefile.btfgen:16]: Shouldn't it be at the top to be checked before the `find` command is executed? It seems to work like this but I don't understand why.

File: Makefile.btfgen [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+else
+Q = @
+msg = @printf '  %-8s %s%s\n' "$(1)" "$(notdir $(2))" "$(if $(3), $(3))";
+MAKEFLAGS += --no-print-directory
+endif
+
+$(BTFHUB_ARCHIVE)/%.btf: $(BTFHUB_ARCHIVE)/%.btf.tar.xz
+	$(call msg,UNTAR,$@)
+	$(Q)tar xvfJ $< -C "$(@D)" > /dev/null
+	$(Q)touch $@
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply Makefile.btfgen]: That's nice. I updated it. Thanks!

File: Makefile [hunk 3]
```diff
@@ -11,6 +11,9 @@ GOHOSTARCH ?= $(shell go env GOHOSTARCH)
 
 ENABLE_BTFGEN ?= false
 
+BPFTOOL ?= bpftool
+ARCH := $(shell uname -m | sed 's/x86_64/x86/' | sed 's/aarch64/arm64/' | sed 's/ppc64le/powerpc/' | sed 's/mips.*/mips/')
```
Comment (alban) [inline Makefile:15]: Should it be changeable by users like this?

```suggestion
ARCH ?= $(shell uname -m | sed 's/x86_64/x86/' | sed 's/aarch64/arm64/' | sed 's/ppc64le/powerpc/' | sed 's/mips.*/mips/')
```

File: .dockerignore [hunk 4]
```diff
@@ -5,5 +5,5 @@
 !gadget-container
 !pkg
 !tools
-!Makefile
+!Makefile*
```
Comment (alban) [inline .dockerignore:8]: Any reason why not calling the file `btfgen.mk`?

File: Makefile.btfgen [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+ifndef ARCH
+$(error ARCH is undefined)
+endif
+
+SOURCE_BTF_FILES = $(shell find $(BTFHUB_ARCHIVE)/ -iregex ".*$(subst x86,x86_64,$(ARCH)).*" -type f -name '*.btf.tar.xz')
+MIN_CORE_BTF_FILES = $(patsubst $(BTFHUB_ARCHIVE)/%.btf.tar.xz, $(OUTPUT)/%.btf, $(SOURCE_BTF_FILES))
+# TODO (filter by architecture too when implemented)
+BPF_O_FILES = $(shell find pkg/gadgets -type f -name '*.o')
+
+all: $(MIN_CORE_BTF_FILES)
```
Comment (alban) [inline Makefile.btfgen:19]:
```suggestion
.PHONY: all
all: $(MIN_CORE_BTF_FILES)
```

File: Makefile.btfgen [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+$(error BTFHUB_ARCHIVE is undefined)
+endif
+
+ifndef ARCH
+$(error ARCH is undefined)
+endif
+
+SOURCE_BTF_FILES = $(shell find $(BTFHUB_ARCHIVE)/ -iregex ".*$(subst x86,x86_64,$(ARCH)).*" -type f -name '*.btf.tar.xz')
+MIN_CORE_BTF_FILES = $(patsubst $(BTFHUB_ARCHIVE)/%.btf.tar.xz, $(OUTPUT)/%.btf, $(SOURCE_BTF_FILES))
+# TODO (filter by architecture too when implemented)
```
Comment (alban) [inline Makefile.btfgen:16]: I think it is fine to fix this TODO in the other PR

**Review by alban — APPROVED** [review]
LGTM

File: Makefile (hunk 3 above)
[the remark this answers is not in the document]
Comment (alban) [reply Makefile:15]: I think it is fine to fix this in the other PR

[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply Makefile]: @eiffel-fl I was iterating again on this to see if BTFGen and arm64 are working fine together. It seems to me they are. I can see the generated BTF files for each platform:

```
$ mkdir -p /tmp/foo/arm64
$ docker create --name arm64 ghcr.io/kinvolk/inspektor-gadget:latest@sha256:02ac6b4aa39b55ab189792ff8b3288304a29ff846aa1494f0b158fac899706a0
$ docker export arm64 > /tmp/foo/arm64/file.tar
$ cd /tmp/foo/arm64
$ tar xvf file.tar
$ tree btfs/ | wc -l
469
```

I opened https://github.com/kinvolk/inspektor-gadget/pull/751 to make it more efficient, but besides that I think it's done. 

What were your concerns about `uname -m` in this case?

**Pull request merged**